File: acm_core.c
/*
 * acm_core.c — ACM-Champernowne core definitions (C implementation)
 *
 * Shot-for-shot remake of acm_core.py. Same functions, same results,
 * different language. The Python and C implementations form two
 * vertices of the DOCS::C::PYTHON audit triangle.
 */

#include "acm_core.h"
#include <math.h>
#include <stdio.h>
#include <string.h>
#include <stdlib.h>


/* -----------------------------------------------------------------
 * n-prime generation
 * ----------------------------------------------------------------- */

int acm_n_primes(int n, int count, int *out)
{
    if (n < 1 || count < 1 || out == NULL)
        return -1;

    if (n == 1) {
        /* Ordinary primes by trial division. */
        int found = 0;
        int candidate = 2;
        while (found < count) {
            int is_prime = 1;
            for (int i = 0; i < found; i++) {
                if (candidate % out[i] == 0) {
                    is_prime = 0;
                    break;
                }
            }
            if (is_prime)
                out[found++] = candidate;
            candidate++;
        }
        return 0;
    }

    /* n >= 2: elements n*k where k is not divisible by n. */
    int found = 0;
    int k = 1;
    while (found < count) {
        if (k % n != 0)
            out[found++] = n * k;
        k++;
    }
    return 0;
}


/* -----------------------------------------------------------------
 * Digit counting (decimal digits in an integer)
 * ----------------------------------------------------------------- */

static int int_digit_count(int x)
{
    if (x <= 0) return 1;
    int d = 0;
    while (x > 0) { d++; x /= 10; }
    return d;
}


/* -----------------------------------------------------------------
 * Champernowne real
 *
 * Precision: the concatenated string is parsed as an IEEE 754 double
 * via strtod(). Only the first ~16 significant decimal digits of the
 * fractional part survive the conversion. For large n or large count,
 * the trailing n-primes contribute nothing to the returned value.
 * This caps effective precision at ~53 bits. The leading-digit
 * property (which needs 1 digit) and the sawtooth structure (which
 * needs ~10) are unaffected.
 * ----------------------------------------------------------------- */

double acm_champernowne_real(int n, int count)
{
    if (n < 1 || count < 1)
        return 0.0;

    int *primes = malloc(count * sizeof(int));
    if (!primes) return 0.0;
    acm_n_primes(n, count, primes);

    /*
     * Build "1." + concatenated decimal representations.
     * Max digits per prime is ~10 (for int range), so a 2048-char
     * buffer is generous for any reasonable count.
     */
    char buf[2048];
    buf[0] = '1';
    buf[1] = '.';
    int pos = 2;

    for (int i = 0; i < count; i++) {
        int written = snprintf(buf + pos, sizeof(buf) - pos, "%d", primes[i]);
        pos += written;
        if (pos >= (int)sizeof(buf) - 1) break;
    }
    buf[pos] = '\0';

    free(primes);
    return strtod(buf, NULL);
}


/* -----------------------------------------------------------------
 * Digit count
 * ----------------------------------------------------------------- */

int acm_digit_count(int n, int count)
{
    if (n < 1 || count < 1)
        return -1;

    int *primes = malloc(count * sizeof(int));
    if (!primes) return -1;
    acm_n_primes(n, count, primes);

    int total = 0;
    for (int i = 0; i < count; i++)
        total += int_digit_count(primes[i]);

    free(primes);
    return total;
}


/* -----------------------------------------------------------------
 * First-digit extraction
 * ----------------------------------------------------------------- */

int acm_first_digit(double x)
{
    if (x <= 0.0)
        return 0;
    double l = log10(x);
    double frac = l - floor(l);
    int d = (int)(pow(10.0, frac) + 1e-9);
    if (d > 9) d = 9;
    return d;
}


/* -----------------------------------------------------------------
 * Benford reference
 *
 * Returns the Benford probability for a single digit d in [1, 9]:
 * log10(1 + 1/d). Takes the digit as an argument and returns a
 * scalar, so callers can request exactly the digit they need
 * without allocating an array. Typical use: compare an observed
 * frequency against acm_benford(d) in a loop over d = 1..9.
 * ----------------------------------------------------------------- */

double acm_benford(int d)
{
    if (d < 1 || d > 9)
        return 0.0;
    return log10(1.0 + 1.0 / d);
}

File: core/acm_core.py
"""
acm_core.py — Central definitions for the ACM-Champernowne process
===================================================================

This module defines the n-prime construction and its derived objects.
All other scripts import from here.

BQN companions (see guidance/BQN-AGENT.md):
    acm_n_primes          -> NPn2 (n >= 2 only)
    acm_champernowne_real -> parse of "1." ++ ChamDigits10
    acm_digit_count       -> DigitCount10
    acm_first_digit       -> LD10 (with +1e-9 guard)
    acm_benford_pmf       -> Benford10

DEFINITIONS
-----------

n-primality:
    A positive integer n >= 1 defining a multiplicative monoid nZ+ = {n, 2n, 3n, ...}.

n-prime:
    An irreducible element of the monoid nZ+.
    The element n*k is n-prime iff n does not divide k.
    Equivalently: multiples of n that are not multiples of n^2.
    For n >= 2, the first n-1 n-primes are n, 2n, 3n, ..., (n-1)n, all below n^2.
    The value n^2 = n*n is the first n-composite (the square boundary).

    For n = 1 (special case), n-primes are ordinary primes.

n-Champernowne real:
    Given the first K n-primes p_1, ..., p_K, concatenate their decimal
    representations after "1." to form a real number.
    Example: 2-primes [2, 6, 10, 14, 18] -> 1.26101418

    The n-Champernowne real encodes both number-theoretic content (primality
    of the monoid) and typographic cost (digit count) in a single scalar.

Decomposition (in log space):
    ln(n-Champernowne) decomposes into:
      - ln(n-primality) = ln(n)         [number-theoretic content, unbounded]
      - ln(n-digitfrac) = ln(digits/K)  [typographic cost, staircase]
      - ln(n-Champernowne)              [full encoding, bounded sawtooth]

    The running mean of the Champernowne reals moves toward ~1.55 = (10+1)/(2*10)
    but we conjecture it approaches from below and never arrives.

CONNECTION TO ARITHMETIC CONGRUENCE MONOIDS (ACMs)
--------------------------------------------------

The monoid nZ+ ∪ {1} is the simplest ACM, denoted M_{n,n} in the literature
(Geroldinger, Halter-Koch, Chapman, Baginski). Hilbert used M_{1,4} = {1,5,9,13,...}
to demonstrate non-unique factorization: 441 = 9 × 49 = 21 × 21.

For n >= 2, unique factorization fails in nZ+. Example in 2Z+:
    36 = 2 × 18 = 6 × 6

The n-Champernowne encoding is new: it turns the irreducible elements of each
ACM into a real number whose significand distribution is exactly uniform (not
Benford), and whose behavior under addition vs. multiplication reveals the
fundamental asymmetry between these operations.

KEY OBSERVATIONS
----------------

1. First-digit distribution of n-Champernowne reals is perfectly uniform (1/9 each),
   NOT Benford. This is the signature of the missing multiplicative structure.

2. Under multiplication, significands converge to Benford within ~10 operations.
   Under addition, significands cycle through digits forever (rolling shutter effect).

3. The sawtooth shape of ln(n-Champernowne) is ln(1+m) where m is the base-10
   mantissa of n. The running mean is the average error of the linear approximation
   — a base-10 cousin of the epsilon correction function ε(m) = log₂(1+m) - m.

4. The rolling shutter shear rate is log₁₀(μ) where μ ≈ 1.55 is the mean of the
   Champernowne reals. This is log₁₀(1 + m̄), the true log evaluated at the mean
   mantissa — the function whose computational cost ε measures.
"""

import numpy as np
from typing import List, Tuple


# ---------------------------------------------------------------------------
# Core generators
# ---------------------------------------------------------------------------

def acm_n_primes(n: int, count: int = 5) -> List[int]:
    """
    Return the first `count` n-primes.

    For n = 1: ordinary primes (trial division).
    For n >= 2: elements n*k where k is not divisible by n.

    These are the irreducible elements of the multiplicative monoid nZ+.
    Returns None on invalid input (n < 1 or count < 1).
    """
    if n < 1 or count < 1:
        return None

    if n == 1:
        primes = []
        candidate = 2
        while len(primes) < count:
            if all(candidate % p != 0 for p in primes):
                primes.append(candidate)
            candidate += 1
        return primes

    result = []
    k = 1
    while len(result) < count:
        if k % n != 0:
            result.append(n * k)
        k += 1
    return result


def acm_champernowne_real(n: int, count: int = 5) -> float:
    """
    Construct the n-Champernowne real from the first `count` n-primes.

    Concatenates the decimal representations of the n-primes after "1."
    Example: n=2, count=5 -> primes [2,6,10,14,18] -> 1.26101418

    Precision: the concatenated string is parsed as an IEEE 754 double
    via float(). Only the first ~16 significant decimal digits of the
    fractional part survive the conversion. For large n or large count,
    the trailing n-primes contribute nothing to the returned value.
    This caps effective precision at ~53 bits. The leading-digit
    property (which needs 1 digit) and the sawtooth structure (which
    needs ~10) are unaffected.

    Returns 0.0 on invalid input.
    """
    primes = acm_n_primes(n, count)
    if primes is None:
        return 0.0
    s = ''.join(str(p) for p in primes)
    return float('1.' + s)


def acm_digit_count(n: int, count: int = 5) -> int:
    """
    Total decimal digits used by the first `count` n-primes.
    This is the "typographic cost" of the Champernowne encoding.
    Returns -1 on invalid input.
    """
    primes = acm_n_primes(n, count)
    if primes is None:
        return -1
    return sum(len(str(p)) for p in primes)


# ---------------------------------------------------------------------------
# Decomposition in log space
# ---------------------------------------------------------------------------

def acm_decompose(n: int, count: int = 5) -> Tuple[float, float, float]:
    """
    Return (ln_champernowne, ln_primality, ln_digitfrac) for a given n.

    ln_champernowne = ln(n-Champernowne real)
    ln_primality    = ln(n)
    ln_digitfrac    = ln(total_digits / count)
    """
    c = acm_champernowne_real(n, count)
    d = acm_digit_count(n, count)
    return (
        np.log(c),
        np.log(n) if n > 0 else 0.0,
        np.log(d / count)
    )


# ---------------------------------------------------------------------------
# Batch computation (numpy convenience — no C equivalent)
# ---------------------------------------------------------------------------

def acm_champernowne_array(n_max: int, count: int = 5) -> np.ndarray:
    """
    Return array of Champernowne reals for n = 1, 2, ..., n_max.
    """
    return np.array([acm_champernowne_real(n, count) for n in range(1, n_max + 1)])


def acm_running_mean(arr: np.ndarray) -> np.ndarray:
    """
    Running (cumulative) mean of an array.
    """
    return np.cumsum(arr) / np.arange(1, len(arr) + 1)


# ---------------------------------------------------------------------------
# First-digit extraction
# ---------------------------------------------------------------------------

def acm_first_digit(x: float) -> int:
    """
    Extract the leading significant digit (1-9) of a positive number.
    Uses log10 to avoid string manipulation.
    Returns 0 for non-positive input.
    """
    if x <= 0:
        return 0
    l = np.log10(x)
    frac = l - np.floor(l)
    return min(int(10**frac + 1e-9), 9)


def acm_first_digit_array(arr: np.ndarray) -> np.ndarray:
    """
    Vectorized first-digit extraction for an array of positive numbers.
    """
    log_arr = np.log10(arr)
    frac = log_arr - np.floor(log_arr)
    return np.minimum((10**frac + 1e-9).astype(int), 9)


# ---------------------------------------------------------------------------
# Benford reference
# ---------------------------------------------------------------------------

def acm_benford_pmf() -> np.ndarray:
    """
    Return Benford's law probabilities for digits 1-9 as a numpy array.

    Computes log10(1 + 1/d) for d = 1..9 and returns all nine values
    at once. This is convenient for vectorized comparisons against
    observed digit distributions: subtract, square, sum, and you have
    a chi-squared-flavored distance from Benford in one expression.
    """
    return np.array([np.log10(1 + 1/d) for d in range(1, 10)])
